fideslib/db/session.py

```diff
@@ -0,0 +1,72 @@
+from __future__ import annotations
+
+import logging
+
+from sqlalchemy import create_engine
+from sqlalchemy.engine import Engine
+from sqlalchemy.engine.url import URL
+from sqlalchemy.orm import Session, sessionmaker
+
+from fideslib.core.config import config
+
+logger = logging.getLogger(__name__)
+
+
+def get_db_engine(database_uri: str | URL | None = None) -> Engine:
+    """
+    Return a database engine.
+
+    If the TESTING environment var is set the database engine returned will
+    be connected to the test DB.
+    """
+    if (
+        not database_uri
+        and not config.database.SQLALCHEMY_DATABASE_URI
+        and not config.database.SQLALCHEMY_TEST_DATABASE_URI
+    ):
+        raise ValueError("No database_uri was provided, or set in the config.")
+
+    uri: str | URL
+
+    if database_uri is None:
+        # Don't override any database_uri explicity passed in
+        if config.is_test_mode:
+            uri = str(config.database.SQLALCHEMY_TEST_DATABASE_URI)
```

Comment:
Won't it get set here? https://github.com/ethyca/fideslib/blob/655c189df6fb33be80809109b3d09b50eb81f288/fideslib/core/config.py#L66

Reply:
Why is the error checking above needed then, if it's guaranteed to be set?
```
    if (
        not database_uri
        and not config.database.SQLALCHEMY_DATABASE_URI
        and not config.database.SQLALCHEMY_TEST_DATABASE_URI
    ):
        raise ValueError("No database_uri was provided, or set in the config.")
```

Reply:
This code was copied, but I suspect whoever did this check did it because mypy doesn't know it can't be None and will complain if you don't check it.

It's also possible I'm wrong and it could be None 😄 . Let me verify.

Reply:
OK, I have confirmed this. From this toml config:

```toml
[database]
SERVER="db"
USER="postgres"
PASSWORD="216f4b49bea5da4f84f05288258471852c3e325cd336821097e1e65ff92b528a"
DB="app"
TEST_DB="test"
```

I get:

```py
from fideslib.core.config import config

print(config.database.SQLALCHEMY_DATABASE_URI)
print(config.database.SQLALCHEMY_TEST_DATABASE_URI)
```

```console
postgresql://postgres:216f4b49bea5da4f84f05288258471852c3e325cd336821097e1e65ff92b528a@db:5432/app
postgresql://postgres:216f4b49bea5da4f84f05288258471852c3e325cd336821097e1e65ff92b528a@db:5432/test
```

So the reason for this check is mypy knows the type is `Optional[PostgresDsn]` but doesn't know about the pydantic validator so it wants you to verify it isn't None.

That said, maybe it's still smart to double check here in case the validators ever change. Something like this?

```py
if database_uri is None:
        # Don't override any database_uri explicitly passed in
        if config.is_test_mode:
            if not config.database.SQLALCHEMY_TEST_DATABASE_URI:
                raise ValueError("No SQLALCHEMY_TEST_DATABASE_URI in the config")
            uri = str(config.database.SQLALCHEMY_TEST_DATABASE_URI)
        else:
            if not config.database.SQLALCHEMY_DATABASE_URI:
                raise ValueError("No SQLALCHEMY_DATABASE_URI in the config")
            uri = str(config.database.SQLALCHEMY_DATABASE_URI)
    else:
        uri = database_uri
```

Reply:
that sounds good to me Paul, thanks for verifying